sapicore/data/sampling/__init__.py

```diff
@@ -0,0 +1,136 @@
+
+    Note
+    ----
+    :meth:`~data.Data.sample` accepts initialized objects of this class.
+    Additional samplers may be implemented in future versions.
+
+    """
+
+    def __init__(self, stratified: bool = False, replace: bool = False):
+        self.stratified = stratified
+        self.replace = replace
+
+    def __call__(self, frame: DataFrame, group_keys: str | list[str], n: int | float):
+        """Draws samples from `frame` after grouping it by `columns`.
+
+        Parameters
+        ----------
+        frame: DataFrame
+            Pandas dataframe containing descriptor labels.
+
+        group_keys: str or list of str
+            Names of descriptors (column headers) to group the dataframe by.
+
+        n: int or float
+            Specifies the number or fraction of samples to draw from each group.
+            While the former is more convenient for equal (disproportionate) sampling and the latter
+            for stratified (proportionate) sampling, integer or float values can be used with either one.
+
+        """
+        # keep track of global indices for subset selection.
+        frame["index"] = frame.index
+        grouped = frame.groupby(group_keys, group_keys=False)
+
+        if self.stratified:
+            # convert `n` to fraction if need be.
+            if isinstance(n, int):
+                frac = len(frame["index"].tolist()) * n
+
+            # perform stratified sampling of `frac` out of every group.
+            subset = grouped.apply(lambda x: x.sample(frac=frac, replace=self.replace))
+
+        else:
+            # convert `n` to integer if need be.
+            if isinstance(n, float):
+                n = int(n * len(frame["index"].tolist()))
+
+            subset = grouped.apply(lambda x: x.sample(n, replace=self.replace))
+
+        return subset["index"].tolist()
+
+
+class CV:
+    """Cross validation base class.
+
+    Parameter
+    ---------
+    data: Data
+        The :class:`data.Data` object for which to generate the cross validator.
+
+    cross_validator: BaseCrossValidator
+        A scikit-learn :class:`BaseCrossValidator` object.
+
+    label_keys: str or list of str, optional
+        One or more descriptor label names. Required for some CV schemes (e.g., for StratifiedKFold, it would be
+        the descriptor whose class label counts should be consistent with their proportions in the full sample).
+
+    group_key: str, optional
+        A grouping descriptor key. Required for some CV schemes (e.g., GroupKFold, for generating batch-like
+        folds with non-overlapping groups w.r.t. some descriptor).
+
+    Note
+    ----
+    Some external libraries extend the cross validator offerings of scikit-learn.
+    This implementation is compatible with any class derived from BaseCrossValidator.
+
+    """
+
+    def __init__(
+        self,
+        data,
+        cross_validator: BaseCrossValidator,
+        label_keys: str | list[str] = None,
```

Comment:
Is the general idea for this and the pandas class above that descriptors will be used to create groups that you want to split the data into and sample from based on those labels? I'm not quite sure I follow that as a general way to split data for train/test!? Wouldn't you have to anyway generate ways of splitting data? If so, why are text labels, that may correspond to specific conditions also correspond to ways of how we split the datasets for train/val etc?

I may be missing some intentions here.

Reply:
Assuming most of this has been clarified in today's discussion. To recap, the strategy for splitting the data into train/test sets often (and in fact typically **should**)  depend on things like the number of samples in each class, where classes are defined by label vectors (in our ecosystem, descriptors). The fact PyTorch doesn't have these features (that sklearn does) is a serious oversight, not an argument against including them.

A dataset may be described by multiple label vectors. Different experiments may require splitting on one/some/all of them, sometimes after sampling with/without replacement, stratified or not. These operations are extremely common, and implementing them from scratch in every project would be wasteful.

`label_keys` are descriptor/metadata **variable** names, not the actual labels. Check out my usage examples and sklearn tutorials to get a deeper sense of the intention here. I avoid opaque implementations as a matter of principle, but there are always exceptions to the rule--these wrappers leverage sklearn in ways I've wanted to for years of doing actual work with it.